control/velocity_controller_lqr/velocity_controller_lqr/velocity_controller_lqr_lib.py

```diff
@@ -0,0 +1,282 @@
+
+        Returns:
+            X: float: roll angle
+            Y: float: pitch angle
+            Z: float: yaw angle
+
+        """
+        y_square = y * y
+
+        t0 = +2.0 * (w * x + y * z)
+        t1 = +1.0 - 2.0 * (x * x + y_square)
+        phi = np.arctan2(t0, t1)
+
+        t2 = +2.0 * (w * y - z * x)
+        t2 = +1.0 if t2 > +1.0 else t2
+        t2 = -1.0 if t2 < -1.0 else t2
+        theta = np.arcsin(t2)
+
+        t3 = +2.0 * (w * z + x * y)
+        t4 = +1.0 - 2.0 * (y_square + z * z)
+        psi = np.arctan2(t3, t4)
+
+        return phi, theta, psi
+
+    @staticmethod
+    def ssa(angle: float) -> float:
+        """Function to convert the angle to the smallest signed angle.
+
+        Args:
+            angle: float: angle in radians
+
+        Returns:
+            angle: float: angle in radians
+
+        """
+        if angle > np.pi:
+            angle -= 2 * np.pi
+        elif angle < -np.pi:
+            angle += 2 * np.pi
+        return angle
+
+    def saturate(self, value: float, windup: bool, limit: float) -> tuple:
+        """Function to saturate the value within the limits, and set the windup flag.
+
+        Args:
+            value: float: value to be saturated
+            windup: bool: windup variable
+            limit: float: limit for saturation
+
+        Returns:
+            (value, windup): tuple: saturated value and windup flag
+
+        """
+        if abs(value) > limit:
+            windup = True
+            value = limit * value / abs(value)
+        else:
+            windup = False
+
+        return value, windup
+
+    @staticmethod
+    def anti_windup(
+        k_i: float, error: float, integral_sum: float, windup: bool
+    ) -> float:
+        """Function to saturate the integral value within the limits.
+
+        Args:
+            k_i : float: integral gain
+            error: float: error value
+            integral_sum: float: integral sum
+            windup: bool: windup variable
+
+
+
+        Returns:
+            value: float: saturated value
+
+        """
+        if windup:
+            integral_sum += 0.0
+        else:
+            integral_sum += k_i * error
+
+        return integral_sum
+
+    def calculate_coriolis_matrix(
+        self, pitch_rate: float, yaw_rate: float, sway: float, heave: float
+    ) -> np.array:
+        """Calculates the 3x3 coriolis matrix.
+
+        Parameters:
+            pitch_rate: float: pitch rate in rad/s
+            yaw_rate: float: yaw rate in rad/s
+            sway: float: sway velocity in m/s
+            heave: float: heave velocity in m/s
+
+        Returns:
+            C: np.array: 3x3 Coriolis Matrix
+
+        """
+        return np.array(
+            [
+                [0.2, -30 * sway * 0.01, -30 * heave * 0.01],
+                [30 * sway * 0.01, 0, 1.629 * pitch_rate],
+                [30 * heave * 0.01, 1.769 * yaw_rate, 0],
+            ]
+        )
+
+    def calculate_lqr_u(
```

Comment:
There is a lot going on in this method. Would be nice if you could split it up into several smaller methods, to enhance readablity

Reply:
This, as in the calculate_lqr_u method